### ISSUE TITLE
StatusBar on Add Card screen and presenting method selection on iPad issues

### ISSUE THREAD

**Issue by DannyRostami** [braintree/braintree-ios-drop-in#419]
### Braintree Drop-in SDK Version

9.8.2

### Environment

Production

### Xcode Version

14.3

### OS Version & Device

any

### Integration type

Swfit Package Manager

### Development Processor

Intel

### Describe the bug

We're having two issues which are the following:
1- StatusBar: We have a screen already presented modal, once we are trying to present drop-in, statusBar on Add Card screen is always light therefore the statusBar not visible on light mode. You can see the screen shot.
2- Presenting method selection: Drop-In presents the method selection screen on iPhone really nicely at the bottom of the screen but on iPad it presents it modally regardless of screen being Slide over or fullscreen.
3- At the end that would be great if we could present Add Card screen modally, currently it presents it in fullscreen.

### To reproduce

Steps to reproduce issue #1:
1- When device is on light appearance.
2- Present Braintree drop-in from a modal screen
3- On method selection choose credit Card or Debit
4- Observe the statusBar not being visible on light mode

Steps to reproduce issue #2:
1- Present Braintree drop-in on iPad in Slide over 
2- Observe the method selection screen is being presented modally and not half like in compact.

### Expected behavior

Expected behaviour for #2 : It should check UserInterfaceSizeClass and not iPad device for presenting method selection.

### Screenshots

| 1-iPad Slide over  | 2-Before presenting Drop-in Add Card screen |  3-After presenting Add Card screen 
| ------------- | ------------- | ------------- | 
| <img src="https://github.com/braintree/braintree-ios-drop-in/assets/119142769/f7bec6d9-bce9-47a5-8fb7-b0e267ad0de4" width="300"> | <img width="300" alt="beforePresentingAddCardScreen" src="https://github.com/braintree/braintree-ios-drop-in/assets/119142769/0ae93266-6c71-433b-98c4-26c66f4126d7">  | <img width="300" alt="addCardStatusBar" src="https://github.com/braintree/braintree-ios-drop-in/assets/119142769/0d9764bb-77a5-4b00-9c2e-9fd655737718">

**Comment by scannillo** [comment]
👋 Hi @DannyRostami - for Issue 2, the UI you're seeing of the centered modal on the iPad is expected as part of our design. That said, the Drop-In could use some UX improvements but our team has been slammed making it lower on our priority list at this moment.

We welcome any and all suggestions you may have for improving the UI and experience! We will gladly take any feedback back to our product teams.

**Comment by scannillo** [comment]
As for Issue 1 - does this only happen when you present Drop-In from a modal?

**Comment by DannyRostami** [comment]
Hi @scannillo thanks for your prompt response, for #1 yeah it happens when presented from a modal screen.

**Comment by DannyRostami** [comment]
I think if Add Card Screen overrides the PreferredStatusBarStyle to .default that would solve the issue.

**Comment by jaxdesmarais** [comment]
Hey @DannyRostami -

I did some testing locally and was able to replicate the behavior you are seeing on iPadOS 16. Interestingly, Apple seems to resolve this in iPadOS 17. I was able to replicate the behavior you are seeing by setting `dropIn.modalPresentationStyle = .fullScreen` in `DemoDropInViewController` in our [Demo app](https://github.com/braintree/braintree-ios-drop-in/tree/main/Demo). 

I was able to resolve this behavior by setting `View controller-based status bar appearance` to `NO` in our Demo apps `Info.plist`. Here is an [Apple thread](https://developer.apple.com/forums/thread/16771) indicating updating your `Info.plist` should resolve the issue. I did not need to make an updates to the `AppDelegate` - the tint color is handled by the SDK so you should only need to make the change to your `plist`. Please do let us know if that setting does not resolve the issue for you.

**Comment by DannyRostami** [comment]
Hi Jax, thanks for the reply. In our project, `View controller-based status bar appearance ` is already set to YES but I have tried it NO and observed Add Card Screen has status bar isn't visible on Light. So as we don't want to set controller-based to NO this solution seems to be not working. 🫠

**Comment by jaxdesmarais** [comment]
Thanks for testing that out @DannyRostami -

Are you able to provide either a small sample app where this is reproducible or preferably the steps we can follow to reproduce this in our Demo app? I've created the branch `modal-demo-app` where you can see the changes applied resolving this issue on a iPad in light mode on the add card screen. My branch may not mimic the set up you are testing on so it'd be helpful to have the code changes we can make to our demo app to reproduce the behavior you are seeing.

**Comment by DannyRostami** [comment]
Thanks for following up on this, the issue is with iPhone status bar not iPad. The iPad issue is different. But on iPhone please follow the following steps:
1- Present a screen modally like screenShot number 2
2- Present drop-in 
3- Select a payment method
4- You will observe add card screen look like screenshot number 3 where you are not able to see status bar

**Comment by DannyRostami** [comment]
Hi @jaxdesmarais is there anything I could provide? I just wanted to say hello and follow on this. ☺️

**Comment by jaxdesmarais** [comment]
Hey @DannyRostami -

Apologies for the delay. I've been testing a few things over the past several days, but the only time I can replicate this is when the phone is set to dark mode and the SDK is set to light mode. Are you able to confirm the phone setting for light vs dark mode when you are reproducing this behavior? Since there are several settings/configurations that go into controlling the status bar color we want to make sure we are replicating this in the same states as when you are seeing this behavior.

**Comment by DannyRostami** [comment]
Hey @jaxdesmarais,
I think I have already stated the settings of the device in above(description and screenshots). The issue will not be seen in dark mode.

**Comment by jaxdesmarais** [comment]
Hello @DannyRostami -

Thanks for confirming, but I am unable to replicate this on iOS 16.4 when setting `dropIn.modalPresentationStyle = .fullScreen` in the `DemoDropInViewController` in our [Demo app](https://github.com/braintree/braintree-ios-drop-in/tree/main/Demo) with both the phone setting and Drop-in settings set to light mode. Here is a screenshot of running the above in our demo app:
| Modal Drop-in |
|-----|
|![simulator_screenshot_A29C3A21-97ED-4FB5-B03D-2F694A716F1F](https://github.com/braintree/braintree-ios-drop-in/assets/20733831/1ec66d49-31ec-49b7-a0be-ba37a64863aa)|

As mentioned previously, are you able to give us the steps to reproduce this in our Demo app? Since we are not seeing the behavior you are seeing, it's quite difficult to say what needs to be updated to get this to behave as you expect.

**Comment by DannyRostami** [comment]
Hi @jaxdesmarais replicating our issue is very easy like I mentioned above all you need is to present drop-in from a modal screen. I have replicated it from your demo app easily. 🤓

https://github.com/braintree/braintree-ios-drop-in/assets/119142769/49d1c679-87d4-46dd-b8e6-738f36f13bdf

**Comment by DannyRostami** [comment]
Also when I have tried to set dropIn.modalPresentationStyle = .fullScreen like you suggested, it fixes the status bar but produces another defect

https://github.com/braintree/braintree-ios-drop-in/assets/119142769/430f193f-313f-44b0-afd7-5bbace4bcd7a

**Comment by DannyRostami** [comment]
I think I have found the solution for it
`[navController setModalPresentationCapturesStatusBarAppearance: true];`

<img width="1004" alt="Screenshot 2023-08-01 at 12 39 46" src="https://github.com/braintree/braintree-ios-drop-in/assets/119142769/2aa7aa23-4629-408d-bc4f-dd5479c01ac2">

**Comment by DannyRostami** [comment]
Hi @jaxdesmarais I was able to find a solution for this issue by using this:
`[navController setModalPresentationCapturesStatusBarAppearance: true];`

<img width="985" alt="Screenshot 2023-08-01 at 13 03 32" src="https://github.com/braintree/braintree-ios-drop-in/assets/119142769/4301a7f9-e8a8-43c5-aed8-74055b9e4891">


Also the UICustomisation in setting should be set to dynamic as well to make it work.

**Comment by DannyRostami** [comment]
@jaxdesmarais It is what I can see from your changes. Still not visible. But we don't want to use .fullscreen it should be modal.

<img width="1210" alt="Screenshot 2023-08-02 at 09 58 16" src="https://github.com/braintree/braintree-ios-drop-in/assets/119142769/660788e8-4afe-49d4-ae90-6088b399337b">

**Comment by StewThom** [comment]
Hi @jaxdesmarais, thanks for spending the time investigating this issue as we know it's fairly minor in the grand scheme of things. 

I took a look at the branch you posted, unfortunately as @DannyRostami mentioned we have a very strict design requirement that the screen preceding the drop in UI needs to be modal and so we cannot use full screen presentation. 

I updated your code to best match our setup as follows  
```
class BlankViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.backgroundColor = .red
    }

    override func viewDidAppear(_ animated: Bool) {
        let vc = FakeCheckoutViewController()
        //  vc.modalPresentationStyle = .fullScreen
        
        self.present(vc, animated: true)
    }
}

class FakeCheckoutViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.backgroundColor = .blue
    }

    override func viewDidAppear(_ animated: Bool) {

	let uiCistomization = BTDropInUICustomization(colorScheme: .dynamic)

	let request = BTDropInRequest()
	request.uiCustomization = uiCistomization
        let dropInController = BTDropInController(
            authorization: "sandbox_9dbg82cq_dcpspy2brwdjr3qn",
            request: request
        ) { controller, result, error in
            // handle result
        }
        
        self.present(dropInController!, animated: true)
    }

}
```

This results in correct behaviour when in dark mode, however the issue still persists in light mode, as can be seen in the following video. 

https://github.com/braintree/braintree-ios-drop-in/assets/6495341/41e87aeb-1054-463a-b8f0-ae2941b47954

When making the change to `BTDropInController`'s `showCardForm` method as @DannyRostami mentioned.

```
- (void)showCardForm:(__unused id)sender {
    BTCardFormViewController* vd = [[BTCardFormViewController alloc] initWithAPIClient:self.apiClient request:self.dropInRequest];
    vd.supportedCardTypes = self.displayCardTypes;
    vd.delegate = self;
    UINavigationController* navController = [[UINavigationController alloc] initWithRootViewController:vd];
    [navController setModalPresentationCapturesStatusBarAppearance: true];
    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad) {
        navController.modalPresentationStyle = UIModalPresentationPageSheet;
    } else {
        navController.modalPresentationStyle = UIModalPresentationOverCurrentContext;
    }
    [self presentViewController:navController animated:YES completion:nil];
}
```

With the addition of `[navController setModalPresentationCapturesStatusBarAppearance: true];` we are seeing expected behaviour as can be observed in this video. 

https://github.com/braintree/braintree-ios-drop-in/assets/6495341/8277b6a6-8d5c-4ed8-9f2f-fcee50ea36e0

Hopefully this helps with your investigations.

**Comment by jaxdesmarais** [comment]
Hello @DannyRostami and @StewThom -

Thanks for getting back to us with those details and your product requirement to present a modal without a `modalPresentationStyle`. 

We need to do some testing with the changes you proposed across a few different presentation styles to ensure this does not cause a regression with other integration patterns. Once we've completed our testing I'll be sure to update you all. Internal ticket for tracking purposes: DTBTSDK-2893

**Comment by jaxdesmarais** [comment]
The PR with this change has been merged - we will let you all know when the fix for this has been released

**Comment by DannyRostami** [comment]
Hi @jaxdesmarais really appreciate your time and considerations. 🤘🏻🤘🏻🤘🏻

**Comment by jaxdesmarais** [comment]
Hello again @DannyRostami and @StewThom -

This has been released in [version 9.9.0](https://github.com/braintree/braintree-ios-drop-in/releases/tag/9.9.0). Thanks again for your patience and raising this issue to us! 🚀
